### PR TITLE
[#9693] refactor isEditAllowed

### LINKED ISSUE #9693: Follow up to 8674 - Basic implementation of `isEditAllowed` is probably not sufficient for some classes
### Bug Description
While working on #8674, I noticed some inconsistencies around `isEditAllowed` specifically the implementation in:

- `CampaignService.java` where a basic implementation was added which needs to be verified if anything else needs to be done
- `CaseService.java` which only contains a basic implementation which is far less sophisticated than the checks for e.g., contacts which looks suspicious.

Investigate the logic and make sure that it is correct and no bugs are introduced by the basic implementation.

### PR DESCRIPTION
**Pull request by JonasCir**
Fixes #9693 
Turns out everything is fine, the checks were just scattered around, I cleaned it up and now everything is just in one place.

### CODE REVIEW

**Review by MateStrysewske — COMMENTED** [review]
I'm a bit doubtful about this PR, honestly. First of all, a method that's called "isEditAllowed" returning anything but a boolean is very counterintuitive to me (I know that this method existed before, but if we're doing a refactoring here, I think renaming the getEditPermissionType methods to this is the wrong way to go). Alternatives could be to only use this method when we're really interested in the ALLOWED enum value and then letting isEditAllowed return a boolean - or to keep the getEditPermissionType name and use it everywhere (and completely get rid of the isEditAllowed name).

Secondly, I don't like replacing the explicit classes with superclasses for the UUID, ARCHIVED and DELETED calls. I know that the IDE suggests those, but @StefanKock has a very good point in saying that using the explicit classes gives more clues as to what exactly we're interested in, and I think that makes a lot of sense. I've added him as another reviewer so that he can say something about it.

Bottom line: I think we should discuss in which way this PR makes sense and whether we might need to change stuff about it.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactJurisdictionFlagsDto.java [hunk 1]
```diff
@@ -20,19 +20,19 @@
 
 public class ContactJurisdictionFlagsDto implements Serializable {
 
-    private Boolean isInJurisdiction;
-    private Boolean isCaseInJurisdiction;
+    private final boolean isInJurisdiction;
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactJurisdictionFlagsDto.java:23]: Why this refactoring? Is it always decidable without null as third option?

File: sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreAdoService.java [hunk 2]
```diff
@@ -99,7 +98,7 @@ public Map<String, Date> calculateEndOfProcessingDate(List<String> entityuuids)
 
 		Expression aggregatedChangeDateExpression = addChangeDates(new AggregatedChangeDateExpressionBuilder(cb), from, true).build();
 		cq.multiselect(from.get(AbstractDomainObject.UUID), cb.max(aggregatedChangeDateExpression));
-		cq.where(from.get(ADO.UUID).in(entityuuids));
+		cq.where(from.get(AbstractDomainObject.UUID).in(entityuuids));
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreAdoService.java:101]: Use `CoreAdo`

**Review by StefanKock — CHANGES_REQUESTED** [review]
I totally agree with @MateStrysewske:
1. `isEditAllowed` not returning a `boolean` but a currently tri-state enum is a lot against naming conventions. I think it should be the other way around: `getEditPermissionType` is valid and `isEditAllowed` has to be removed or changed to `return getEditPermissionType == EditPermissionType.ALLOWED;` (as convenience, but is still slightly duplicate).
2. Changes of UUID / ARCHIVED constants to superclasses should always reflect the class that you expect to handle. In `AbstractCoreAdoService` changing it all to `CoreAdo` (instead of `Case`) is reasonable and clean code. But don't change it in `CampaignService` since it makes clear that it's the Campaigns attributes.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactJurisdictionFlagsDto.java (hunk 1 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactJurisdictionFlagsDto.java:23]: removed from PR.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreAdoService.java (hunk 2 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreAdoService.java:101]: done

**Comment by JonasCir** [comment]
@MateStrysewske @Totally agree.

So I updated the PR quite a bit which gives as good wins w.r.t. readability and LOC.

1. `isEditAllowed` returns a boolean as it should, making the call sites much cleaner.
2. I refactored the edit checks in S2S in such a way that they do not rely on calls into subclasses anymore
3. I introduced the `AbstractEditAllowedDetailView` which removes quite some code duplication from the UI w.r.t. edit checks

File: sormas-ui/src/main/java/de/symeda/sormas/ui/samples/AbstractSampleView.java [hunk 3]
```diff
@@ -118,15 +118,15 @@ protected void setSubComponent(DirtyStateComponent newComponent) {
...
-		if (!isSampleEditAllowed) {
+		Boolean isSampleEditAllowed = isEditAllowed();
+		if (Boolean.FALSE.equals(isSampleEditAllowed)) {
 			component.setEnabled(false);
 		}
 	}
 
-	protected Boolean isSampleEditAllowed() {
-		return FacadeProvider.getSampleFacade().isSampleEditAllowed(getSampleRef().getUuid());
+	protected Boolean isEditAllowed() {
```
Comment (StefanKock) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/samples/AbstractSampleView.java:127]: Why keep Object Boolean here?

File: sormas-backend/src/main/java/de/symeda/sormas/backend/sample/SampleService.java [hunk 4]
```diff
@@ -1094,7 +1094,8 @@ public Predicate createDefaultFilter(CriteriaBuilder cb, Root<Sample> root) {
 		return cb.isFalse(root.get(Sample.DELETED));
 	}
 
-	public Boolean isSampleEditAllowed(Sample sample) {
+
+	public Boolean isEditAllowed(Sample sample) {
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/sample/SampleService.java:1098]: This should not be an Object Boolean, and the current implementation does not support it. 
`inJurisdictionOrOwned(sample).getInJurisdiction()::Boolean` will run into NullPointerException.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractDetailView.java [hunk 5]
```diff
@@ -115,4 +115,5 @@ protected R getReference() {
 	 *            The URL parameters String
 	 */
 	protected abstract void initView(String params);
+
```
Comment (StefanKock) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractDetailView.java:118]: Why empty line here? It's against code convention

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractEditAllowedDetailView.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * In addition to that, it contains generic code to check if the object is editable.
+ *
+ * @param <R>
+ *            {@link ReferenceDto} with the uuid as parsed from the URL.
+ */
+public abstract class AbstractEditAllowedDetailView<R extends ReferenceDto> extends AbstractDetailView<R> {
+
+	private CoreFacade coreFacade;
+
+	protected AbstractEditAllowedDetailView(String viewName, CoreFacade coreFacade) {
```
Comment (StefanKock) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractEditAllowedDetailView.java:19]: This pattern caches the Facade instance which is discouraged because it does not refresh when the sormas-ear is redeployed.

**Review by StefanKock — CHANGES_REQUESTED** [review]
The approach with usage of `isEditAllowed` and `getEditPermissionType` looks now way cleaner. But some new remarks.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreFacadeEjb.java [hunk 7]
```diff
@@ -259,8 +259,13 @@ public Date calculateEndOfProcessingDate(String entityUuid) {
 		return service.calculateEndOfProcessingDate(Collections.singletonList(entityUuid)).get(entityUuid);
 	}
 
-	public EditPermissionType isEditAllowed(String uuid) {
-		ADO ado = service.getByUuid(uuid);
-		return service.isEditAllowed(ado);
+	@Override
+	public EditPermissionType getEditPermissionType(String uuid) {
```
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreFacadeEjb.java:263]: Good point, thanks. I think this is a fundamental utility method which should be callable unconditionally.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/samples/AbstractSampleView.java (hunk 3 above)
Comment (JonasCir) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/samples/AbstractSampleView.java:127]: changed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/sample/SampleService.java (hunk 4 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/sample/SampleService.java:1098]: fixed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractEditAllowedDetailView.java (hunk 6 above)
Comment (JonasCir) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractEditAllowedDetailView.java:19]: fixed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractDetailView.java (hunk 5 above)
Comment (JonasCir) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractDetailView.java]: fixed

**Comment by MateStrysewske** [comment]
@JonasCir Code looks good to me, can you resolve the merge conflict please?
@StefanKock Can you do another review?

**Comment by JonasCir** [comment]
conflict resolved

**Comment by MateStrysewske** [comment]
I'll have to postpone testing this because it's blocked by https://github.com/hzi-braunschweig/SORMAS-Project/issues/8541.

**Review by MateStrysewske — APPROVED** [review]

**Comment by JonasCir** [comment]
@MateStrysewske Thx, can we merge or should we wait for a review from @StefanKock ? :)

**Comment by MateStrysewske** [comment]
Wait for his approval please :)

**Review by StefanKock — APPROVED** [review]
Thx for waiting ;-) I have no objections.

**Comment by JonasCir** [comment]
All good, thank you both! :rocket:
